ffmpeg/lpms_ffmpeg.c

```diff
@@ -650,68 +705,46 @@ int process_out(struct input_ctx *ictx, struct output_ctx *octx, AVCodecContext
-      if (AVMEDIA_TYPE_VIDEO == ost->codecpar->codec_type &&
-        inf && inf->hw_frames_ctx && filter->hwframes &&
-        inf->hw_frames_ctx->data != filter->hwframes) {
-      free_filter(&octx->vf); // XXX really should flush filter first
-      ret = init_video_filters(ictx, octx);
-      if (ret < 0) return lpms_ERR_FILTERS;
-    }
-    frame = filter->frame;
-    av_frame_unref(frame);
-    ret = av_buffersrc_write_frame(filter->src_ctx, inf);
-    if (ret < 0) proc_err("Error feeding the filtergraph\n");
-    ret = av_buffersink_get_frame(filter->sink_ctx, frame);
-    frame->pict_type = AV_PICTURE_TYPE_NONE;
-    if (AVERROR(EAGAIN) == ret || AVERROR_EOF == ret) frame = NULL;
-    else if (ret < 0) proc_err("Error consuming the filtergraph");
-    tb = av_buffersink_get_time_base(filter->sink_ctx);
-  } else frame = inf;
-
-  // encode
-  av_init_packet(&pkt);
-  if (encoder) {
-    if (frame || !inf) {
-      // only send if we've received a frame from filtergraph or this is a flush
-      ret = avcodec_send_frame(encoder, frame);
-      if (AVERROR_EOF == ret) ;
-      else if (ret < 0) proc_err("Error sending frame to encoder\n");
-    }
-    ret = avcodec_receive_packet(encoder, &pkt);
-    if (AVERROR(EAGAIN) == ret || AVERROR_EOF == ret) return ret;
-    if (ret < 0) proc_err("Error receiving packet from encoder\n");
-    tb = encoder->time_base;
-  } else proc_err("Trying to transmux") // XXX pass in the inpacket, set  pkt = ipkt
 
+  if (!encoder) proc_err("Trying to transmux; not supported")
 
-  // packet bookkeeping.  XXX use av_rescale_delta for audio
-  pkt.stream_index = ost->index;
-  if (av_cmp_q(tb, ost->time_base)) {
-    pkt.pts = av_rescale_q_rnd(pkt.pts, tb, ost->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
-    pkt.dts = av_rescale_q_rnd(pkt.dts, tb, ost->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
-    pkt.duration = av_rescale_q(pkt.duration, encoder->time_base, ost->time_base);
+  if (!filter || !filter->active) {
+    // No filter in between decoder and encoder, so use input frame directly
+    ret = encode(encoder, inf, octx, ost);
+    if (ret < 0) return ret;
   }
 
-  // drop any preroll audio. may need to drop multiple packets for multichannel
-  // XXX this breaks if preroll isn't exactly one AVPacket or drop_ts == 0
-  //     hasn't been a problem in practice (so far)
-  if (AVMEDIA_TYPE_AUDIO == ost->codecpar->codec_type) {
-      if (octx->drop_ts == AV_NOPTS_VALUE) octx->drop_ts = pkt.pts;
-      if (pkt.pts && pkt.pts == octx->drop_ts) goto proc_cleanup;
+  // Sometimes we have to reset the filter if the HW context is updated
+  // because we initially set the filter before the decoder is fully ready
+  // and the decoder may change HW params
+  if (AVMEDIA_TYPE_VIDEO == ost->codecpar->codec_type &&
+      inf && inf->hw_frames_ctx && filter->hwframes &&
+      inf->hw_frames_ctx->data != filter->hwframes) {
+    free_filter(&octx->vf); // XXX really should flush filter first
```

Comment:
is this some tech debt that we should write a ticket for and take care of later?   or might it be something small that can be added to this PR?

Reply:
If this is referring to the "really should flush filter first" comment -- Not entirely sure TBH. The reason I'm hesitant to fix this [1] or make a issue is mostly because I haven't yet looked at how to construct a test case to trigger breakage and verify the fix. Without a breaking case, it may not even be a legitimate concern. (I just spent a few minutes trying some things without luck.)

[1] Which should be simpler to fix now with this changeset. Most everything between the `while(1)` could be moved into a standalone function and invoked just prior to freeing the filters. But again, I don't want to blindly add code without demonstrating actual breakage.